:art: new styleguide (#37)
ZestCommunity/ZestCode: docs/STYLEGUIDE.md

Reviewer: I didn't see this in the styleguide, but the styleguide should encourage the usage of fixed width integer types.

Author: added

> # PROS Kernel Styleguide

Reviewer 2: Shouldn't this say ZestCode?

> [!TIP]
> ZestCode is in the process of migrating to this new style, some older code may not yet be using this new style.

> Maintaining a consistent coding style is important. It helps make the code more readable. ZestCode's style guide is largely inspired by [rustfmt](https://doc.rust-lang.org/nightly/style-guide/).

Reviewer: ```suggestion
Maintaining a consistent coding style is important. It helps make the code more readable. ZestCode's style guide is largely inspired by [rustfmt](https://doc.rust-lang.org/stable/style-guide/).
```
If this wasn't intentional, the stable channel's link should probably be used.

ZestCode uses [clang-format](https://releases.llvm.org/19.1.0/tools/clang/docs/ClangFormat.html) to format its source code.
We recommend using [clangd](https://marketplace.visualstudio.com/items?itemName=llvm-vs-code-extensions.vscode-clangd) to automatically format your code while you are working on it.
Some contributors have set up VSCode so it format their code whenever they save or make a commit.

Some additional notes follow:

## File extensions

> [!IMPORTANT]
> ZestCode is being rewritten in C++. No C files or headers may be added, but existing ones can be modified.

- C source files will have the extension `.c`
- C++ source files will have the extension `.cpp`
- C header files will have the extension `.h`
- C++ header files will have the extension `.hpp`
- Assembly source files will have the extension `.s`
- markdown documents will have the extension `.md`

## Naming Conventions

- **files, folders** - snake_case
- **classes, structs, enums, unions, concepts, etc** - PascalCase
- **functions** - snake_case
- **variables** - snake_case
- **member variable** - m_snake_case (not a typo)
- **macros** - SCREAMING_SNAKE_CASE
- **compile-time constants** - SCREAMING_SNAKE_CASE

> - **enum members** - SCREAMING_SNAKE_CASE

Reviewer 3: this is probably fine for raw C enums, but we can do snake case for enum classes

Author: fixed

Reviewer 3: I meant that we don't have to do the screaming part for enum classes, and we should specify raw C enums vs C++ enum classes having different styling.

# ZestCode Styleguide

> [!TIP]

> > ZestCode is in the process of migrating to this new style, some older code may not yet be using this new style.

Reviewer: ```suggestion
> ZestCode is in the process of migrating to this new style; some older code may not yet be using this new style.
```
Grammar nitpick

Maintaining a consistent coding style is important. It helps make the code more readable. ZestCode's style guide is largely inspired by [rustfmt](https://doc.rust-lang.org/stable/style-guide/).

## General Rules

> [!TIP]

> > [clangd](https://marketplace.visualstudio.com/items?itemName=llvm-vs-code-extensions.vscode-clangd) can be used to automatically format your code using the ZestCode styleguide

Reviewer: ```suggestion
> The [clangd VSCode extension](https://marketplace.visualstudio.com/items?itemName=llvm-vs-code-extensions.vscode-clangd) can be used to automatically format your code using the ZestCode styleguide
```
Bit pedantic, but the link doesn't link to clangd, it links to a clangd *VSCode extension*.

Some basic rules are as follows:

- indent using 4 spaces, not tabs
- no more than 100 columns per line
- write tests for your code
- always use `enum class` over `enum`

> - a `struct` should only have member variables, and no access modifiers

Reviewer: ```suggestion
- a `struct` should only be used for an object where all fields/methods are public
```

Author: I don't agree with this.

Methods should be part of classes, not structs. Structs should be PODs (plain old datatypes), as they have guarantees that no longer exist as soon as we a constructor for the struct.

- don't use `typedef`

> - use [C++11 attributes](https://en.cppreference.com/w/cpp/language/attributes)

Reviewer: ```suggestion
- use [C++11 attributes](https://en.cppreference.com/w/cpp/language/attributes) instead of the older GCC attribute syntax: `__attribute__((...))`
```

- only use macros as a last resort

For more details, see [.clang-format](../.clang-format).

## File extensions

> [!IMPORTANT]
> ZestCode is being rewritten in C++. No C files or headers may be added, but existing ones can be modified.

- C source files will have the extension `.c`
- C++ source files will have the extension `.cpp`
- C header files will have the extension `.h`
- C++ header files will have the extension `.hpp`

> - Assembly source files will have the extension `.s`

Reviewer: If we add assembly extensions, we should add `.py` and `.ld`, since they show up more or as often as assembly files in the kernel.

- markdown documents will have the extension `.md`

## Naming Conventions

- **files, folders** - snake_case
- **classes, structs, enums, unions, concepts, etc** - PascalCase
- **functions** - snake_case
- **variables** - snake_case

> - **member variable** - m_snake_case (not a typo)

Reviewer: ```suggestion
- **member variables** - m_snake_case (The `m_` is intentional)
```
Also, all the example names should be inside ``s so that they render as code, since they are technically example code samples.

- **macros** - SCREAMING_SNAKE_CASE
- **compile-time constants** - SCREAMING_SNAKE_CASE
- **enum members** - snake_case

## Documentation

Use Doxygen-style comments as shown in the sections below.

> ### File-Level Comments

Reviewer: Are we keeping the file-level comments?

Author: we shouldn't, I'll remove it

```c++
/**
 * @brief Brief description of the function.
 *
 * An extended description of the function (if applicable).
 *
 * This function uses the following values of errno when an error state is
 * reached:
 * ERRNO_VALUE - Description of what causes this error
 *
 * @tparam T the type of the parameters
 *
 * @param parameter_name The parameter description
 * @param other_parameter_name The parameter description
 *
 * @return The description of the return value, if this is longer than one line
 * then it will wrap around under the return statement
 *
 * @code
 * // example code here
 * @endcode
 *
 * @see src/referenced_file.cpp
 */
```

### Inline Implementation Comments

Sometimes it is necessary to explain a particularly complex statement or series of statements. In this case, you should use inline comments, placed either immediately before or trailing the line or lines in question. In general, prefer placing such comments before offending lines, unless the comment is quite short. These comments should start with a `//` followed by a space. If they are placed trailing a line, they should be separated from the end of the line by one space.

```c++

> float Q_rsqrt(float number) {

Author: fixed

And for the record, I wasn't the one who put it here lol

        E_MEMBER_1, // these can be omitted if it's painfully obvious
        E_MEMBER_2, // what each is for, or if there are just so many of
        E_MEMBER_3, // them it doesn't make practical sense to

>         E_MEMBER_4 // document them all.

Reviewer 3: my understanding is that enums are supposed to be canonically this for doxygen:
```
E_MEMBER_EXAMPLE, ///< Description here about enum
```

Author: fixed

}
```

> [!TIP]
> In the above example, the comments describing each member of the `enum` run together and form complete sentences for effect.
> Please do not do this in your code!

### Function-Level Comments

These should be placed immediately before the function prototype they are describing in a header file.

```c++
/**
 * @brief Brief description of the function.
 *
 * An extended description of the function (if applicable).
 *
 * This function uses the following values of errno when an error state is
 * reached:
 * ERRNO_VALUE - Description of what causes this error
 *
 * @param parameter_name The parameter description
 * @param other_parameter_name The parameter description
 *
 * @return The description of the return value, if this is longer than one line
 * then it will wrap around under the return statement

>  */

Reviewer 2: Should probably add `@tparam`, as well as `@see` for referenced files imo

Author: done

    return y;
}
```
> [!TIP]
> In the above example, there is a line of code that has been commented out.
> This is fine to do while testing, but any commented out lines of code should be removed before any merge into the master branch takes place, unless a compelling reason can be presented for them to remain.

> All that being said, try to avoid code that is so complex that it requires inline comments for its purpose to be clear.

Reviewer: ```suggestion
Keep inline comments to a minimum. When possible, avoid writing complex code that can't be understood without inline comments.
```
Rather ironic that a sentence on clarity is itself hard to read.

#### Notes to Other Developers (Or Yourself)

When writing code, it can sometimes be useful to leave notes to other developers or to yourself in the future. Examples of these include:

- `// TODO: something that should be done`
- `// HACK: used to describe a particularly nasty way of solving a problem-- could be improved, but it works for now`

> While it is not strictly necessary to use these keywords in comments, they can be helpful-- modern editors either highlight some of these keywords by default or have extensions that do.

Reviewer: ```suggestion
While it is not strictly necessary to use these keywords in comments, they can be helpful—modern editors either highlight some of these keywords by default or have extensions that do.
```
This should use an em dash.
